cpp/src/arrow/dataset/scanner.cc

```diff
@@ -428,6 +428,11 @@ class ARROW_DS_EXPORT AsyncScanner : public Scanner,
   Result<FragmentGenerator> GetFragments() const;
 
   std::shared_ptr<Dataset> dataset_;
+
+  // XXX if Scanner were truly single-use this would be a 1:1 relationship
+  std::unordered_map<std::shared_ptr<compute::ExecPlan>,
+                     std::shared_ptr<compute::ExecContext>>
+      plans_;
```

Comment:
Well I originally did this because the destruction of ExecPlan was waiting for the plan to finish, blocking the only thread which could finish the plan. However removing it doesn't seem to reproduce the deadlock so I guess `stop_producing` is a sufficient keepalive.